File: backend/openclaw/ubuntu/qwen.js
const fs = require("fs");
const path = require("path");
const { callUbuntuHealth, callUbuntuOpenClaw } = require("./qwen_process");
const {
  handleRagRequest,
  resolveUseRagFlag,
} = require("../rag/rag_process");

function parseJsonSafely(rawBody) {
  if (typeof rawBody !== "string" || !rawBody.trim()) return {};
  try {
    return JSON.parse(rawBody);
  } catch {
    return {};
  }
}

function normalizeOptionalObject(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? value : {};
}

function parseBooleanLike(value, fallback = false) {
  if (typeof value === "boolean") return value;
  if (typeof value === "number") return value !== 0;
  if (typeof value === "string") {
    const normalized = value.trim().toLowerCase();
    if (["1", "true", "yes", "y", "on"].includes(normalized)) return true;
    if (["0", "false", "no", "n", "off"].includes(normalized)) return false;
  }
  return fallback;
}

function writeQwenLog(entry) {
  try {
    const logPath = path.join(__dirname, "..", "api_log", "request_input.json");
    fs.mkdirSync(path.dirname(logPath), { recursive: true });
    fs.writeFileSync(logPath, JSON.stringify(entry, null, 2), "utf8");
  } catch {
    // Ignore logging failures.
  }
}

function parseMaybeNumber(value) {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string" && value.trim()) {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : undefined;
  }
  return undefined;
}

function parseMaybeBoolean(value, fallback = false) {
  if (typeof value === "boolean") return value;
  if (typeof value === "string") {
    const normalized = value.trim().toLowerCase();
    if (["1", "true", "yes", "y", "on"].includes(normalized)) return true;
    if (["0", "false", "no", "n", "off"].includes(normalized)) return false;
  }
  if (typeof value === "number") return value !== 0;
  return fallback;
}

function normalizeMessageContent(content) {
  if (typeof content === "string") return content.trim();

  if (Array.isArray(content)) {
    return content
      .map((part) => {
        if (typeof part === "string") return part.trim();
        if (part && typeof part === "object") {
          if (typeof part.text === "string") return part.text.trim();
          if (typeof part.content === "string") return part.content.trim();
        }
        return "";
      })
      .filter(Boolean)
      .join(" ");
  }

  return "";
}

function buildAgentMessages(payload) {
  const messages = [];
  const systemParts = [];

  if (typeof payload.system_context === "string" && payload.system_context.trim()) {
    systemParts.push(payload.system_context.trim());
  }

  const scenarioText =
    typeof payload.scenario === "string" && payload.scenario.trim()
      ? payload.scenario.trim()
      : "";
  if (scenarioText) systemParts.push(`scenario: ${scenarioText}`);

  const userState = normalizeOptionalObject(payload.user_state);
  if (Object.keys(userState).length > 0) systemParts.push(`user_state: ${JSON.stringify(userState)}`);

  const metadata = normalizeOptionalObject(payload.metadata);
  if (Object.keys(metadata).length > 0) systemParts.push(`metadata: ${JSON.stringify(metadata)}`);

  if (systemParts.length > 0) {
    messages.push({ role: "system", content: systemParts.join("\n\n") });
  }

  const history = Array.isArray(payload.history) ? payload.history : [];
  for (const item of history) {
    if (!item) continue;
    if (typeof item === "string") {
      const content = item.trim();
      if (content) messages.push({ role: "user", content });
      continue;
    }
    if (typeof item !== "object") continue;

    const rawRole =
      typeof item.role === "string" && item.role.trim()
        ? item.role.trim().toLowerCase()
        : "user";
    const content = normalizeMessageContent(item.content);
    if (!content) continue;
    const role = ["user", "assistant", "tool"].includes(rawRole) ? rawRole : "user";
    messages.push({ role, content });
  }

  const userMessage =
    (typeof payload.user_message === "string" && payload.user_message.trim()
      ? payload.user_message.trim()
      : "") ||
    (typeof payload.message === "string" && payload.message.trim()
      ? payload.message.trim()
      : "") ||
    (typeof payload.content === "string" && payload.content.trim()
      ? payload.content.trim()
      : "");

  if (userMessage) messages.push({ role: "user", content: userMessage });
  if (messages.length === 0) messages.push({ role: "user", content: "xin chao" });
  return messages;
}

function buildQwenRequestPayload({ payload, config, modelName, userKey, sessionKey, useLora, normalizedOptions }) {
  const temperature = parseMaybeNumber(payload.temperature) ?? parseMaybeNumber(config.temperature) ?? 0.7;
  const maxTokens = parseMaybeNumber(payload.max_tokens) ?? parseMaybeNumber(config.max_tokens) ?? 256;
  const enableThinking = parseMaybeBoolean(
    payload.enable_thinking,
    parseMaybeBoolean(config.enable_thinking, false),
  );

  return {
    session_id: payload.session_id,
    turn_id: payload.turn_id,
    system_context: typeof payload.system_context === "string" ? payload.system_context : "",
    user_message: payload.user_message,
    history: Array.isArray(payload.history) ? payload.history : [],
    scenario: payload.scenario,
    user_state: normalizeOptionalObject(payload.user_state),
    runtime: normalizeOptionalObject(payload.runtime),
    options: normalizedOptions,
    metadata: normalizeOptionalObject(payload.metadata),
    model: modelName,
    user: userKey,
    session_key: sessionKey,
    messages: buildAgentMessages(payload),
    reasoning_effort: payload.reasoning_effort,
    temperature,
    max_tokens: maxTokens,
    top_p: payload.top_p,
    enable_thinking: enableThinking,
    use_lora: useLora,
    use_rag: resolveUseRagFlag(payload, false),
    rag_top_k: payload.rag_top_k,
    model_code: typeof payload.model_code === "string" ? payload.model_code : undefined,
    x_openclaw_model:
      typeof payload.x_openclaw_model === "string" ? payload.x_openclaw_model : undefined,
    ubuntu_model: typeof payload.ubuntu_model === "string" ? payload.ubuntu_model : undefined,
  };
}

function createUbuntuRoute({ sendJson, sendError, readBody, config = {} }) {
  return async function ubuntuRoute(req, res, url, rawBody = "") {
    if (!url.pathname.startsWith("/openclaw/")) return false;

    if (url.pathname === "/openclaw/health" || url.pathname === "/openclaw/v1/health") {
      if (req.method !== "GET") return sendError(res, 405, "Method not allowed");
      try {
        const health = await callUbuntuHealth(config);
        return sendJson(res, 200, { ok: true, route: url.pathname, brand: "openclaw", ubuntu: health });
      } catch (error) {
        return sendError(res, 502, error.message || "Ubuntu health check failed");
      }
    }

    if (url.pathname === "/openclaw/agent/health") {
      if (req.method !== "GET") return sendError(res, 405, "Method not allowed");
      return sendJson(res, 200, {
        ok: true,
        route: url.pathname,
        brand: "openclaw",
        runtime: "qwen-agent",
      });
    }

    if (url.pathname === "/openclaw/agent/level") {
      if (req.method !== "POST") return false;

      const payload = parseJsonSafely(rawBody || (await readBody(req)));
      writeQwenLog({
        timestamp: new Date().toISOString(),
        path: url.pathname,
        method: req.method,
        payload,
      });

      return false;
    }

    if (
      url.pathname !== "/openclaw/chat/completions" &&
      url.pathname !== "/openclaw/v1/chat/completions" &&
      url.pathname !== "/openclaw/agent/run"
    ) {
      return false;
    }

    if (req.method !== "POST") return sendError(res, 405, "Method not allowed");

    try {
      const payload = parseJsonSafely(rawBody || (await readBody(req)));
      if (!payload || typeof payload !== "object") {
        return sendError(res, 400, "JSON body must be an object");
      }

      const runtime = normalizeOptionalObject(payload.runtime);
      const modelName =
        (typeof runtime.model === "string" && runtime.model.trim()) ||
        (typeof payload.model === "string" && payload.model.trim()) ||
        (typeof config.model === "string" && config.model.trim()) ||
        (typeof config.ubuntu_model === "string" && config.ubuntu_model.trim()) ||
        (typeof config.x_openclaw_model === "string" && config.x_openclaw_model.trim()) ||
        "Qwen3.5-4B-V4";
      const userKey =
        (typeof payload.user === "string" && payload.user.trim()) ||
        (typeof payload.user_id === "string" && payload.user_id.trim()) ||
        (typeof payload.session_id === "string" && payload.session_id.trim()) ||
        "media_tech:request";
      const sessionKey =
        (typeof payload.session_key === "string" && payload.session_key.trim()) ||
        (typeof payload.session_id === "string" && payload.session_id.trim()) ||
        "agent:media_tech:ubuntu";
      const useLora = typeof payload.use_lora === "boolean" ? payload.use_lora : (typeof config.use_lora === "boolean" ? config.use_lora : true);
      const normalizedOptions = normalizeOptionalObject(payload.options);

      writeQwenLog({
        timestamp: new Date().toISOString(),
        path: url.pathname,
        method: req.method,
        payload,
      });

      const requestPayload = buildQwenRequestPayload({
        payload,
        config,
        modelName,
        userKey,
        sessionKey,
        useLora,
        normalizedOptions,
      });

      if (requestPayload.use_rag) {
        const ragResponse = await handleRagRequest({
          requestPayload,
          config,
          stage: "run",
        });
        if (ragResponse) {
          writeQwenLog({
            timestamp: new Date().toISOString(),
            path: url.pathname,
            method: req.method,
            payload,
            use_rag_handled: true,
          });
          return sendJson(res, 200, ragResponse);
        }
      }

      const result = await callUbuntuOpenClaw({ requestPayload, config, stage: "run" });
      return sendJson(res, 200, result);
    } catch (error) {
      return sendError(res, 500, error.message || "Qwen GPU websocket route failed");
    }
  };
}

module.exports = {
  buildAgentMessages,
  buildQwenRequestPayload,
  createUbuntuRoute,
};

File: backend/openclaw/ubuntu/qwen.py
#!/usr/bin/env python3
import json
import os
import sys
from copy import deepcopy
from typing import Any, Dict, List


DEFAULT_SYSTEM_MESSAGE = (
    "Bạn là trợ lý tư vấn dịch vụ/quán/sản phẩm chạy sau OpenClaw. "
    "Trả lời tiếng Việt ngắn gọn, rõ ý."
)


def _parse_json(raw: str):
    if not isinstance(raw, str) or not raw.strip():
        return {}
    try:
        raw = raw.lstrip("\ufeff")
        parsed = json.loads(raw)
        return parsed if isinstance(parsed, dict) else {}
    except Exception:
        return {}


def _read_stdin_text() -> str:
    raw = sys.stdin.buffer.read()
    if not raw:
        return ""
    for encoding in ("utf-8-sig", "utf-8", "cp1258", "cp1252", "latin-1"):
        try:
            return raw.decode(encoding)
        except Exception:
            continue
    return raw.decode("utf-8", errors="replace")


def _write_stdout_json(data: Dict[str, Any]) -> None:
    text = json.dumps(data, ensure_ascii=False)
    sys.stdout.buffer.write(text.encode("utf-8"))
    sys.stdout.buffer.flush()


def _stringify_content(content: Any) -> str:
    if content is None:
        return ""

    if isinstance(content, str):
        return content.strip()

    if isinstance(content, list):
        parts = []
        for item in content:
            if isinstance(item, dict):
                text = item.get("text")
                if isinstance(text, str) and text.strip():
                    parts.append(text.strip())
                else:
                    parts.append(json.dumps(item, ensure_ascii=False))
            else:
                text = str(item).strip()
                if text:
                    parts.append(text)
        return " ".join(parts).strip()

    if isinstance(content, dict):
        return json.dumps(content, ensure_ascii=False)

    return str(content).strip()


def _normalize_message(message):
    if not isinstance(message, dict):
        return None

    role = str(message.get("role", "user")).strip().lower()
    if role not in {"system", "user", "assistant", "tool"}:
        role = "user"

    content = _stringify_content(message.get("content", ""))
    if not content:
        return None

    normalized = {
        "role": role,
        "content": content,
    }

    for key in ("name", "tool_call_id", "id"):
        value = message.get(key)
        if isinstance(value, str) and value.strip():
            normalized[key] = value.strip()

    function_call = message.get("function_call")
    if isinstance(function_call, dict):
        fn_name = function_call.get("name")
        fn_args = function_call.get("arguments")
        if isinstance(fn_name, str) and fn_name.strip():
            normalized["function_call"] = {
                "name": fn_name.strip(),
                "arguments": _stringify_content(fn_args),
            }

    return normalized


def _normalize_history(history):
    if not isinstance(history, list):
        return []

    normalized = []
    for item in history:
        entry = _normalize_message(item)
        if entry:
            normalized.append(entry)

    return normalized


def _first_nonempty(*values: Any, default: Any = "") -> Any:
    for value in values:
        if isinstance(value, str):
            if value.strip():
                return value.strip()
        elif value not in (None, "", [], {}, ()):
            return value
    return default


def _parse_number(value: Any) -> Any:
    if isinstance(value, bool):
        return None

    if isinstance(value, (int, float)):
        return value

    if isinstance(value, str) and value.strip():
        try:
            parsed = float(value)
        except Exception:
            return None
        return int(parsed) if parsed.is_integer() else parsed

    return None


def _parse_bool(value: Any, default: bool = False) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return bool(value)
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized in {"1", "true", "yes", "y", "on"}:
            return True
        if normalized in {"0", "false", "no", "n", "off"}:
            return False
    return default


def _default_options(options: Dict[str, Any]) -> Dict[str, Any]:
    current = dict(options) if isinstance(options, dict) else {}

    current.setdefault("mode", "run")
    current.setdefault("need_decision", True)
    current.setdefault("allow_tool_call", True)
    current.setdefault("max_history", 20)

    return current


def _ensure_first_system_message(messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    if not messages:
        return [
            {"role": "system", "content": DEFAULT_SYSTEM_MESSAGE},
            {"role": "user", "content": "xin chào"},
        ]

    has_system = any(m.get("role") == "system" for m in messages)

    if not has_system:
        return [{"role": "system", "content": DEFAULT_SYSTEM_MESSAGE}] + messages

    if messages[0].get("role") == "system":
        return messages

    system_messages = [m for m in messages if m.get("role") == "system"]
    non_system_messages = [m for m in messages if m.get("role") != "system"]

    merged_system_content = "\n\n".join(
        _stringify_content(m.get("content", ""))
        for m in system_messages
        if _stringify_content(m.get("content", ""))
    ).strip()

    if not merged_system_content:
        merged_system_content = DEFAULT_SYSTEM_MESSAGE

    return [{"role": "system", "content": merged_system_content}] + non_system_messages


def _default_messages_from_payload(data: Dict[str, Any]) -> List[Dict[str, Any]]:
    messages: List[Dict[str, Any]] = []

    raw_messages = data.get("messages")

    # Ưu tiên messages do JS backend đã build sẵn.
    # Nếu JS đã đưa system_context vào role system thì Python giữ nguyên.
    if isinstance(raw_messages, list) and raw_messages:
        for item in raw_messages:
            normalized = _normalize_message(item)
            if normalized:
                messages.append(normalized)

    else:
        # Fallback ổn định:
        # Nếu JS chưa build messages nhưng có system_context,
        # Python tự đưa system_context vào role system.
        system_context = _stringify_content(data.get("system_context"))
        if system_context:
            messages.append({
                "role": "system",
                "content": system_context,
            })

        history = _normalize_history(data.get("history"))
        if history:
            messages.extend(history)

        user_message = _stringify_content(data.get("user_message"))
        if user_message:
            messages.append({
                "role": "user",
                "content": user_message,
            })

    if not messages:
        messages = [{"role": "user", "content": "xin chào"}]

    return _ensure_first_system_message(messages)


def normalize_payload(payload):
    data = deepcopy(payload) if isinstance(payload, dict) else {}

    runtime = data.get("runtime") if isinstance(data.get("runtime"), dict) else {}
    metadata = data.get("metadata") if isinstance(data.get("metadata"), dict) else {}
    options = data.get("options") if isinstance(data.get("options"), dict) else {}
    config = data.get("config") if isinstance(data.get("config"), dict) else {}

    model = str(
        _first_nonempty(
            runtime.get("model"),
            data.get("model"),
            data.get("upstream_model"),
            data.get("ubuntu_model"),
            os.getenv("QWEN_MODEL_NAME"),
            default="Qwen3.5-4B-V3",
        )
    ).strip()

    user = str(data.get("user", "")).strip() or "media_tech:request"
    session_key = str(data.get("session_key", "")).strip() or "agent:media_tech:ubuntu"

    temperature = _first_nonempty(
        _parse_number(data.get("temperature")),
        _parse_number(config.get("temperature")),
        _parse_number(os.getenv("QWEN_TEMPERATURE")),
        default=0.4,
    )

    max_tokens = _first_nonempty(
        _parse_number(data.get("max_tokens")),
        _parse_number(config.get("max_tokens")),
        _parse_number(os.getenv("QWEN_MAX_TOKENS")),
        default=128,
    )

    enable_thinking = _parse_bool(
        data.get("enable_thinking"),
        _parse_bool(config.get("enable_thinking"), False),
    )

    normalized = {
        "model": model,
        "user": user,
        "session_key": session_key,
        "messages": _default_messages_from_payload(data),
        "reasoning_effort": str(data.get("reasoning_effort", "minimal")).strip() or "minimal",
        "temperature": temperature,
        "max_tokens": max_tokens,
        "top_p": data.get("top_p", 0.9),
        "enable_thinking": enable_thinking,
        "use_lora": bool(data.get("use_lora", True)),
        "use_rag": bool(data.get("use_rag", False)),
        "rag_top_k": data.get("rag_top_k", 4),
        "options": _default_options(options),
        "normalized": True,
    }

    # Giữ lại system_context để backend/server sau còn debug hoặc forward nếu cần.
    for key in ("session_id", "turn_id", "user_message", "scenario", "system_context"):
        value = data.get(key)
        if isinstance(value, str) and value.strip():
            normalized[key] = value.strip()

    for key in ("user_state", "runtime", "metadata"):
        value = data.get(key)
        if isinstance(value, dict):
            normalized[key] = value

    if "history" in data:
        normalized["history"] = _normalize_history(data.get("history"))

    model_server = _first_nonempty(
        data.get("model_server"),
        runtime.get("model_server"),
        metadata.get("model_server"),
        os.getenv("QWEN_MODEL_SERVER"),
        default="",
    )
    if isinstance(model_server, str) and model_server.strip():
        normalized["model_server"] = model_server.strip()

    for key in (
        "x_openclaw_model",
        "ubuntu_model",
        "message_channel",
        "upstream_model",
    ):
        value = data.get(key)
        if isinstance(value, str) and value.strip():
            normalized[key] = value.strip()

    return normalized


def _build_base_output(normalized: Dict[str, Any], error_reason: str = "") -> Dict[str, Any]:
    base = deepcopy(normalized) if isinstance(normalized, dict) else {}

    if not base.get("messages"):
        base["messages"] = [
            {"role": "system", "content": DEFAULT_SYSTEM_MESSAGE},
            {"role": "user", "content": "xin chào"},
        ]
    else:
        base["messages"] = _ensure_first_system_message(base["messages"])

    base["options"] = _default_options(
        base.get("options") if isinstance(base.get("options"), dict) else {}
    )

    base["model"] = base.get("model") or "Qwen3.5-4B-V3"
    base["user"] = base.get("user") or "media_tech:request"
    base["session_key"] = base.get("session_key") or "agent:media_tech:ubuntu"
    base["temperature"] = base.get("temperature", 0.4)
    base["max_tokens"] = base.get("max_tokens", 128)
    base["top_p"] = base.get("top_p", 0.9)
    base["enable_thinking"] = bool(base.get("enable_thinking", False))
    base["use_lora"] = bool(base.get("use_lora", True))
    base["use_rag"] = bool(base.get("use_rag", False))
    base["rag_top_k"] = base.get("rag_top_k", 4)
    base["normalized"] = bool(base.get("normalized", False))

    if error_reason:
        base["normalized"] = False
        base["formatter_reason"] = error_reason

    return base


def main():
    try:
        raw = _read_stdin_text()
        payload = _parse_json(raw)
        normalized = normalize_payload(payload)
        result = _build_base_output(normalized)

    except Exception as ex:
        result = {
            "normalized": False,
            "error": str(ex),
            "model": "Qwen3.5-4B-V3",
            "user": "media_tech:request",
            "session_key": "agent:media_tech:ubuntu",
            "messages": [
                {"role": "system", "content": DEFAULT_SYSTEM_MESSAGE},
                {"role": "user", "content": "xin chào"},
            ],
            "options": _default_options({}),
            "temperature": _parse_number(os.getenv("QWEN_TEMPERATURE")) or 0.4,
            "max_tokens": _parse_number(os.getenv("QWEN_MAX_TOKENS")) or 128,
            "top_p": 0.9,
            "enable_thinking": _parse_bool(os.getenv("QWEN_ENABLE_THINKING"), False),
            "use_lora": True,
            "use_rag": False,
            "rag_top_k": 4,
        }

    _write_stdout_json(result)


if __name__ == "__main__":
    main()
